### ISSUE TITLE
Extend StatsBase.model_response methods with a Distribution argument

### ISSUE THREAD

**Issue by dmbates** [JuliaStats/StatsModels.jl#27]
I have been thinking about conversion of a binary `PooledDataArray` or `CategoricalVector` to a 0/1 floating point vector in the `StatsBase.model_response` method.  The trick is in determining what kind of response vector (or matrix or other structure) is desired.  It occurred to me that the Distribution type in the model tells us that.  For an `lm` or, in my case, `lmm` model the distribution type is Normal.  For a `glm` or `glmm` model the distribution type is given in the model specification.

In creating a `GLM.GlmResp` object the response vector is checked to ensure that all values are in the support of the distribution so we might as well use that as the guide.

If this seems reasonable I will create a pull request.

Is it worthwhile changing the name of this extractor at the same time?  The name `model_response` was copied from R and is not Julianesque.

**Comment by ararslan** [comment]
That definitely sounds reasonable to me! It sounds like it could be ported easily to the statistical modeling functionality that remains in DataFrames as well (though I can take care of that).

I agree that `model_response` is not a great name. What would you suggest as an alternative?

**Comment by dmbates** [comment]
Perhaps just `response`

**Comment by ararslan** [comment]
Sounds good to me.

**Comment by kleinschmidt** [comment]
How about storing the contrasts for the response along with the contrasts for predictor variables in the `ModelFrame`?

**Comment by ararslan** [comment]
That seems reasonable

**Comment by dmbates** [comment]
I don't think this is quite the same as storing contrasts.  Contrasts determine the conversion of categorical covariates into a set of columns in a model matrix.  There are no constraints on the values in those columns except that they are of the element type of the model matrix.

Conversion of a categorical response to a response vector can be different.  The most common case is a binary categorical response to the support of a Bernoulli random variable (i.e. 0/1 in whatever element type is indicated).   In that case, the only question is which value should be mapped to zero and which to one.  I would say that the default should be the order of the levels so the code looks like
```
T[cv == levels(catvec)[2] for cv in catvec]
```
where `catvec` is the categorical vector and `T` is the desired element type.  If the user wants the other ordering they should use an `OrderedCategoricalVector` to achieve it.

How about starting small, with just this type of conversion and seeing what others suggest themselves as we go along?

**Comment by dmbates** [comment]
Presently `StatsModels` does not use `Distributions`.  Is having `StatsModels` depend on `Distribtutions` too high a price to pay for this?

**Comment by dmbates** [comment]
See #28

**Comment by nalimilan** [comment]
Contrasts could make sense to determine what's the reference category. But that's quite different from the contrasts we currently have, so maybe that's not a good idea.

Using the distributions from Distributions.jl is appealing, but it introduces a dependency on Rmath at the moment. That said, GLM and MixedModels also depend on it, so that's probably OK.
